lib/Sema/TypeCheckStmt.cpp

```diff
@@ -806,31 +808,54 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
         if (S->getTargetName() == (*I)->getLabelInfo().Name) {
           Target = *I;
           break;
+        } else {
+          unsigned distance =
+            TC.getCallEditDistance(S->getTargetName(), (*I)->getLabelInfo().Name,
+                                   TypeChecker::UnreasonableCallEditDistance);
+          if (distance < TypeChecker::UnreasonableCallEditDistance)
+            labelCorrections.insert(distance, std::move(*I));
         }
       }
+      labelCorrections.filterMaxScoreRange(
+        TypeChecker::MaxCallEditDistanceFromBestCandidate);
     }
     
     if (!Target) {
       // If we're in a defer, produce a tailored diagnostic.
       if (isInDefer()) {
         TC.diagnose(S->getLoc(), diag::jump_out_of_defer, "break");
-        return nullptr;
-      }
-      
-      auto diagid = diag::break_outside_loop;
-
-      // If someone is using an unlabeled break inside of an 'if' or 'do'
-      // statement, produce a more specific error.
-      if (S->getTargetName().empty() &&
-          std::any_of(ActiveLabeledStmts.rbegin(),
-                      ActiveLabeledStmts.rend(),
-                      [&](Stmt *S) -> bool {
-                        return isa<IfStmt>(S) || isa<DoStmt>(S);
-                      })) {
-        diagid = diag::unlabeled_break_outside_loop;
+      } else if (S->getTargetName().empty()) {
+        // If we're dealing with an unlabeled break inside of an 'if' or 'do'
+        // statement, produce a more specific error.
+        if (std::any_of(ActiveLabeledStmts.rbegin(),
+                        ActiveLabeledStmts.rend(),
+                        [&](Stmt *S) -> bool {
+                          return isa<IfStmt>(S) || isa<DoStmt>(S);
+                        })) {
+          TC.diagnose(S->getLoc(), diag::unlabeled_break_outside_loop);
+        } else {
+          // Otherwise produce a generic error.
+          TC.diagnose(S->getLoc(), diag::break_outside_loop);
+        }
+      } else {
+        // If an unresolved label was used, but we have a single correction,
+        // produce the specific diagnostic and fixit.
+        if (labelCorrections.size() == 1) {
+          for (auto &entry : labelCorrections)
+            TC.diagnose(S->getTargetLoc(), diag::unresolved_label_corrected,
+                        S->getTargetName(), (*entry.Value).getLabelInfo().Name)
+              .fixItReplace(SourceRange(S->getTargetLoc()),
+                            (*entry.Value).getLabelInfo().Name.get());
+        } else {
+          // If we have multiple corrections or none, produce a generic diagnostic
+          // and all fixits available.
+          auto diag = TC.diagnose(S->getTargetLoc(), diag::unresolved_label,
+                                  S->getTargetName());
+          for (auto &entry : labelCorrections)
+            diag.fixItReplace(SourceRange(S->getTargetLoc()),
+                              (*entry.Value).getLabelInfo().Name.get());
```

Comment:
Fix-its aren't alternatives; they're *all* applied. You'll have to either use multiple notes (in which case you could point at the statement that defines each label), or just string-append all the names into the diagnostic. (Or use a clever `%select` to put multiple variations in one diagnostic: "'foo'", "'foo' or 'bar'", "'foo', 'bar', or 'baz'".)

Reply:
Yes, I tried to model on identifier typo correction, but I clearly misread that code on my first pass. Thank you for pointing this out.
I also refactored the shared logic into a static method.